core/cache/controller_test.go

```diff
@@ -224,17 +213,51 @@ func (s *ControllerSuite) TestRemoveUnit(c *gc.C) {
 	s.AssertResident(c, unit.CacheId(), false)
 }
 
+func (s *ControllerSuite) TestMarkAndSweep(c *gc.C) {
+	controller, events := s.new(c)
+
+	// Note that the model change is processed last.
+	s.processChange(c, charmChange, events)
+	s.processChange(c, appChange, events)
+	s.processChange(c, machineChange, events)
+	s.processChange(c, unitChange, events)
+	s.processChange(c, modelChange, events)
+
+	done := make(chan struct{})
+	go func() {
+		// Removals are congruent with FILO.
+		// Model is last because models are added if they do not exist,
+		// when we first get a delta for one of their entities.
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveUnit{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveMachine{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveApplication{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveCharm{})
+		c.Check(s.nextChange(c, events), gc.FitsTypeOf, cache.RemoveModel{})
+		done <- struct{}{}
```

Comment:
these are the bare channel reads that are dangerous.
If you just care about a single done event then you can just "close(done)" and you don't have to worry about blocking.

However, imagine if one of the following asserts fail (say line 239). In that case, the TestMarkAndSweep exits with a panic, and it will never call <-done to unblock this goroutine.